pkg/genericapiserver/server/serve_test.go

```diff
@@ -564,6 +566,312 @@ NextTest:
+	validateIpLimitState(t, &ipLimit, "Second increment call for ipBasedLimit", 1, 2, 2, 3)
+	val, ok = ipLimit.limits["1.2.3.4"]
+	assert.True(t, ok, "Increment call for ipBasedLimit had no val")
+	assert.EqualValues(t, val, 2, "Post increment ipBasedLimit has a max of %d", val)
+
+	err = ipLimit.incrementIp("4.3.2.1")
+	assert.NoError(t, err)
+	validateIpLimitState(t, &ipLimit, "First increment call for second ip ipBasedLimit", 2, 2, 3, 3)
+	val, ok = ipLimit.limits["4.3.2.1"]
+	assert.True(t, ok, "Increment call for ipBasedLimit had no val")
+	assert.EqualValues(t, val, 1, "Post increment ipBasedLimit has a max of %d", val)
+
+	err = ipLimit.incrementIp("4.3.2.1")
+	assert.Error(t, err)
+	if limErr, ok := err.(limitedError); ok {
+		assert.True(t, limErr.Timeout(), "IpLimit limitedError should return timeout true")
+		assert.True(t, limErr.Temporary(), "IpLimit limitedError should return temporary true")
+	} else {
+		assert.Fail(t, "Expected a limitedError from incrementIp fail but got %T", err)
+	}
+	validateIpLimitState(t, &ipLimit, "Second increment call for ipBasedLimit", 2, 2, 3, 3)
+	val, ok = ipLimit.limits["4.3.2.1"]
+	assert.True(t, ok, "Increment call for ipBasedLimit had no val")
+	assert.EqualValues(t, val, 1, "Post increment ipBasedLimit has a max of %d", val)
+
+	ipLimit.decrementIp("1.2.3.4")
+	validateIpLimitState(t, &ipLimit, "Decrement call for ipBasedLimit", 2, 2, 2, 3)
+	val, ok = ipLimit.limits["1.2.3.4"]
+	assert.True(t, ok, "Decrement call for ipBasedLimit had no val")
+	assert.EqualValues(t, val, 1, "Post increment ipBasedLimit has a max of %d", val)
+
+	ipLimit.decrementIp("1.2.3.4")
+	validateIpLimitState(t, &ipLimit, "Deleting decrement call for ipBasedLimit", 1, 2, 1, 3)
+	val, ok = ipLimit.limits["1.2.3.4"]
+	assert.False(t, ok, "Deleting decrement call for ipBasedLimit had a val")
+	assert.EqualValues(t, val, 0, "Deleting decrement call for ipBasedLimit has a val of %d", 0)
+	val, ok = ipLimit.limits["4.3.2.1"]
+	assert.True(t, ok, "Unmodified IP for ipBasedLimit had no val")
+	assert.EqualValues(t, val, 1, "Unmodified IP for ipBasedLimit has a max of %d", val)
+
+	ipLimit.decrementIp("4.3.2.1")
+	validateIpLimitState(t, &ipLimit, "Deleting decrement call for ipBasedLimit", 0, 2, 0, 3)
+	val, ok = ipLimit.limits["4.3.2.1"]
+	assert.False(t, ok, "Deleting decrement call for ipBasedLimit had a val")
+	assert.EqualValues(t, val, 0, "Deleting decrement call for ipBasedLimit has a val of %d", 0)
+}
+
+func TestParrallelIpLimitAccessNoLimit(t *testing.T) {
+	ipLimit := ipBasedLimit{limits: make(map[string]int), ipMax: 10, max: 100}
+	stop := time.Now().Add(time.Second * 10) // Long enough to prevent flaky ???
+	stateHolder := parrallelStateHolder{ipLimit: &ipLimit, waitSize: 100,
+		stop: stop, good: 0, limit: 0, count: 0}
+	ips := []string{"1.1.1.1", "2.2.2.2", "3.3.3.3", "4.4.4.4", "5.5.5.5",
+		"6.6.6.6", "7.7.7.7", "8.8.8.8", "9.9.9.9",
+		"1973:dead:beef:0121:1973:dead:beef:0121"}
+	done := make(chan struct{})
```

Comment:
Insert a `,100`-- as written (unbuffered), the channel is acting to serialize all operations done on it.